hail/src/main/scala/is/hail/expr/types/physical/PSubsetStruct.scala

```diff
@@ -0,0 +1,182 @@
+    ps.fieldOffset(structAddress, idxMap(fieldIdx))
+
+  override def fieldOffset(structAddress: Code[Long], fieldIdx: Int): Code[Long] =
+    ps.fieldOffset(structAddress, idxMap(fieldIdx))
+
+  def loadField(structAddress: Code[Long], fieldName: String): Code[Long] =
+    ps.loadField(structAddress, fieldName)
+
+  def loadField(structAddress: Long, fieldName: String): Long =
+    ps.loadField(structAddress, fieldName)
+
+  override def loadField(structAddress: Long, fieldIdx: Int): Long =
+    ps.loadField(structAddress, idxMap(fieldIdx))
+
+  override def loadField(structAddress: Code[Long], fieldIdx: Int): Code[Long] =
+    ps.loadField(structAddress, idxMap(fieldIdx))
+
+  override def setFieldPresent(structAddress: Code[Long], fieldName: String): Code[Unit] = ???
+
+  override def setFieldMissing(structAddress: Code[Long], fieldName: String): Code[Unit] = ???
+
+  override def setFieldMissing(structAddress: Long, fieldIdx: Int): Unit = ???
+
+  override def setFieldMissing(structAddress: Code[Long], fieldIdx: Int): Code[Unit] = ???
+
+  override def setFieldPresent(structAddress: Long, fieldIdx: Int): Unit = ???
+
+  override def setFieldPresent(structAddress: Code[Long], fieldIdx: Int): Code[Unit] = ???
+
+  def insertFields(fieldsToInsert: TraversableOnce[(String, PType)]): PSubsetStruct = ???
+
+  override def initialize(structAddress: Long, setMissing: Boolean): Unit = ???
+
+  override def stagedInitialize(structAddress: Code[Long], setMissing: Boolean): Code[Unit] = ???
+
+  override def allocate(region: Region): Long = ???
+
+  override def allocate(region: Code[Region]): Code[Long] = ???
+
+  override def setRequired(required: Boolean): PType = ???
+
+  override def copyFromType(mb: EmitMethodBuilder[_], region: Value[Region], srcPType: PType, srcAddress: Code[Long], deepCopy: Boolean): Code[Long] = ???
+
+  override def copyFromTypeAndStackValue(mb: EmitMethodBuilder[_], region: Value[Region], srcPType: PType, stackValue: Code[_], deepCopy: Boolean): Code[_] = ???
+
+  override protected def _copyFromAddress(region: Region, srcPType: PType, srcAddress: Long, deepCopy: Boolean): Long = ???
+
+  override def constructAtAddress(mb: EmitMethodBuilder[_], addr: Code[Long], region: Value[Region], srcPType: PType, srcAddress: Code[Long], deepCopy: Boolean): Code[Unit] = ???
+
+  override def constructAtAddress(addr: Long, region: Region, srcPType: PType, srcAddress: Long, deepCopy: Boolean): Unit = ???
+}
+
+object PSubsetStructSettable {
+  def apply(cb: EmitCodeBuilder, pt: PStruct, name: String, sb: SettableBuilder): PSubsetStructSettable = {
+    new PSubsetStructSettable(pt, sb.newSettable(name))
+  }
+}
+
+class PSubsetStructSettable(val pt: PStruct, val a: Settable[Long]) extends PBaseStructValue with PSettable {
+  def get: PSubsetStructCode = new PSubsetStructCode(pt, a)
+
+  // Again, is fieldIdx here the backing PStruct?
+  def loadField(cb: EmitCodeBuilder, fieldIdx: Int): IEmitCode = {
```

Comment:
if `pt` isn't PSubsetStruct, then I can't use the PSubsetStruct.loadField, which means that the fieldIdx must correspond to the underlying PStruct, not the index in the view.

Reply:
I don't think PSubsetStructSettable should take a PStruct and a `a` to construct -- it should take a PStructSettable and the PSubsetStruct. That will let you clean this up for sure.

Reply:
Ok, I must have misread your last comment asking for this change. I had this taking PSubsetStruct in the initial revision of the PR(though Settable, not PSubsetStructSettable)

Reply:
I don't fully understand the comment about PStructSettable. The 2nd argument (a) needs to be a Settable, and it can't take itself. I'm not sure I fully see the problem in the current / original design (is it related to the comment? If so I think that's sorted: the fieldIdx in loadField should be called on the PSubsetStruct, and the value should be the address of the full struct in region. PSubsetStruct needs to understand how to perform read operations on fields of that full struct, that is its purpose). 

Could you clarify if the current design doesn't meet your goals?